File: docs/sphinx/examples/cpp/providers/nvcf_state.cpp
// Compile and run with:
// ```
// nvq++ --target nvcf --nvcf-backend tensornet nvcf_state.cpp -o out.x &&
// ./out.x
// ```
// Assumes a valid NVCF API key and function ID have been stored in environment
// variables or `~/.nvcf_config file`. Alternatively, they can be set in the
// command line like below.
// ```
// nvq++ --target nvcf --nvcf-backend tensornet --nvcf-api-key <YOUR API KEY>
// --nvcf-function-id <NVCF function Id> nvcf_state.cpp -o out.x && ./out.x
// ```
// Please refer to the documentations for information about how to attain NVCF
// information.

#include "cudaq/algorithms/state.h"
#include <cudaq.h>
#include <iostream>

int main() {
  auto kernel = cudaq::make_kernel();
  const std::size_t NUM_QUBITS = 20;
  auto q = kernel.qalloc(NUM_QUBITS);
  kernel.h(q[0]);
  for (std::size_t qId = 0; qId < NUM_QUBITS - 1; ++qId)
    kernel.x<cudaq::ctrl>(q[qId], q[qId + 1]);
  auto state = cudaq::get_state(kernel);
  std::cout << "Amplitude(00..00) = " << state[0] << "\n";
  std::cout << "Amplitude(11..11) = " << state[(1ULL << NUM_QUBITS) - 1]
            << "\n";
}

File: pr-1189/sphinx/examples/python/providers/nvcf_state.py
import cudaq

# This example assumes the NVCF API key and Function Id have been set in the `~/.nvcf_config` file/environment variables.
# If not, you can set the API Key and Function ID environment variables in the Python script with:
# ```
# os.environ["NVCF_API_KEY"] = "<YOUR NVCF API KEY>"`
# os.environ["NVCF_FUNCTION_ID"] = "<YOUR NVCF FUNCTION ID>"
# ```
# Alternatively, the `api_key` and `function_id` values can be passed to the target directly,
# ```
# cudaq.set_target("nvcf",
#                 api_key="<YOUR NVCF API KEY>"
#                 function_id="<YOUR NVCF FUNCTION ID>")
# ```
cudaq.set_target("nvcf")

num_qubits = 20
kernel = cudaq.make_kernel()
qubits = kernel.qalloc(num_qubits)
# Place qubits in GHZ state.
kernel.h(qubits[0])
for i in range(num_qubits - 1):
    kernel.cx(qubits[i], qubits[i + 1])

state = cudaq.get_state(kernel)
print("Amplitude(00..00) =", state[0])
print("Amplitude(11..11) =", state[2**num_qubits - 1])
